src/architecture/unixfs.md

```diff
@@ -0,0 +1,611 @@
+  the path components used for sharding. It MUST be `murmur3-x64-64` (`0x22`).
+- `decode(PBNode.Data).Data.Data` is a bit field, which indicates whether or not
+  links are part of this HAMT, or its leaves. The usage of this field is unknown, given
+  that you can deduce the same information from the link names.
+- `decode(PBNode.Data).Data.fanout` MUST be a power of two. This encodes the number
+  of hash permutations that will be used on each resolution step. The log base 2
+  of the `fanout` indicate how wide the bitmask will be on the hash at for that step.
+  `fanout` MUST be between 8 and probably 65536. <!-- 65536 is a totally arbitrary choice I made, FIXME: get consensus on an upper bound. -->.
+
+The field `Name` of an element of `PBNode.Links` for a HAMT starts with an
+uppercase hex-encoded prefix, which is `log2(fanout)` bits wide.
+
+##### Path Resolution
+
+To resolve the path inside a HAMT:
+
+1. Take the current path component, then hash it using the [multihash] represented
+   by the value of `decode(PBNode.Data).hashType`.
+2. Pop the `log2(fanout)` lowest bits from the path component hash digest, then
+   hex encode (using 0-F) those bits using little endian. Find the link that starts
+   with this hex encoded path.
+3. If the link `Name` is exactly as long as the hex encoded representation, follow
+   the link and repeat step 2 with the child node and the remaining bit stack.
+   The child node MUST be a HAMT directory, or else the directory is invalid. Otherwise, continue.
+4. Compare the remaining part of the last name you found. If it matches the original
+   name you were trying to resolve, you have successfully resolved a path component.
+   Everything past the hex encoded prefix is the name of that element, which is useful when listing children of this directory.
+
+### `TSize` / `DagSize`
+
+This is an optional field in `PBNode.Links[]`. It **does not** represent any
+meaningful information of the underlying structure, and there is no known
+usage of it to this day, although some implementations omit these.
+
+To compute the `DagSize` of a node, which is stored in the parents, sum the length of the `dag-pb` outside message binary length and the `blocksizes` of all child files.
+
+An example of where this could be useful is as a hint to smart download clients.
+For example, if you are downloading a file concurrently from two sources that have
+radically different speeds, it would probably be more efficient to download bigger
+links from the fastest source, and smaller ones from the slowest source.
+
+<!--TODO: check that this is true-->
+There is no failure mode known for this field, so your implementation should be
+able to decode nodes where this field is wrong (not the value you expect), or 
+partially or completely missing. This also allows smarter encoders to give a
+more accurate picture (Don't count duplicate blocks, etc.).
+
+### Metadata
+
+UnixFS currently supports two optional metadata fields.
```

Comment:
what is 'supports' - what expectations are there around generation / parsing of these? - we have implementations that don't encode mode fully